File: src/lib/chunker.ts
export interface TextChunk {
  text: string;
  index: number;
  pageNumber: number;
  metadata: {
    documentId: string;
    startChar: number;
    endChar: number;
  };
}

interface ChunkerOptions {
  chunkSize?: number;
  chunkOverlap?: number;
}

const DEFAULT_CHUNK_SIZE = 512;
const DEFAULT_CHUNK_OVERLAP = 50;

/**
 * Splits text into overlapping chunks using a recursive character splitter approach.
 * Tries to split on paragraph boundaries first, then sentences, then words.
 */
export function chunkText(
  text: string,
  documentId: string,
  pageNumber: number,
  options: ChunkerOptions = {}
): TextChunk[] {
  const chunkSize = options.chunkSize ?? DEFAULT_CHUNK_SIZE;
  const chunkOverlap = options.chunkOverlap ?? DEFAULT_CHUNK_OVERLAP;

  const cleanedText = cleanText(text);

  if (cleanedText.length <= chunkSize) {
    return [
      {
        text: cleanedText,
        index: 0,
        pageNumber,
        metadata: { documentId, startChar: 0, endChar: cleanedText.length },
      },
    ];
  }

  const chunks: TextChunk[] = [];
  let start = 0;
  let chunkIndex = 0;

  while (start < cleanedText.length) {
    let end = start + chunkSize;

    if (end < cleanedText.length) {
      // Try to break at a paragraph boundary
      const paragraphBreak = cleanedText.lastIndexOf("\n\n", end);
      if (paragraphBreak > start + chunkSize * 0.5) {
        end = paragraphBreak;
      } else {
        // Try to break at a sentence boundary
        const sentenceBreak = findLastSentenceBreak(cleanedText, start, end);
        if (sentenceBreak > start + chunkSize * 0.3) {
          end = sentenceBreak;
        } else {
          // Try to break at a word boundary
          const wordBreak = cleanedText.lastIndexOf(" ", end);
          if (wordBreak > start) {
            end = wordBreak;
          }
        }
      }
    } else {
      end = cleanedText.length;
    }

    const chunkText = cleanedText.slice(start, end).trim();
    if (chunkText.length > 0) {
      chunks.push({
        text: chunkText,
        index: chunkIndex,
        pageNumber,
        metadata: { documentId, startChar: start, endChar: end },
      });
      chunkIndex++;
    }

    start = end - chunkOverlap;
    if (start >= cleanedText.length) break;
  }

  return chunks;
}

/**
 * Chunks a multi-page document where pages are separated.
 */
export function chunkDocument(
  pages: { text: string; pageNumber: number }[],
  documentId: string,
  options: ChunkerOptions = {}
): TextChunk[] {
  const allChunks: TextChunk[] = [];
  let globalIndex = 0;

  for (const page of pages) {
    const pageChunks = chunkText(page.text, documentId, page.pageNumber, options);
    for (const chunk of pageChunks) {
      allChunks.push({ ...chunk, index: globalIndex });
      globalIndex++;
    }
  }

  return allChunks;
}

function cleanText(text: string): string {
  return text
    .replace(/\r\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/[ \t]+/g, " ")
    .trim();
}

function findLastSentenceBreak(
  text: string,
  start: number,
  end: number
): number {
  const sentenceEnders = [".", "!", "?", "。", "！", "？"];
  let lastBreak = -1;

  for (let i = end; i > start; i--) {
    if (sentenceEnders.includes(text[i]) && i + 1 < text.length && text[i + 1] === " ") {
      lastBreak = i + 1;
      break;
    }
  }

  return lastBreak;
}

File: backend/app/services/chunker.py
"""Text chunking with recursive splitting on natural boundaries."""

import re
from dataclasses import dataclass

from app.config import settings


@dataclass
class TextChunk:
    text: str
    index: int
    page_number: int
    document_id: str
    start_char: int
    end_char: int


def clean_text(text: str) -> str:
    text = text.replace("\r\n", "\n")
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r"[ \t]+", " ", text)
    return text.strip()


def find_last_sentence_break(text: str, start: int, end: int) -> int:
    """Find the last sentence-ending punctuation followed by a space."""
    sentence_enders = ".!?。！？"
    for i in range(end, start, -1):
        if i < len(text) and text[i] in sentence_enders:
            if i + 1 < len(text) and text[i + 1] == " ":
                return i + 1
    return -1


def chunk_text(
    text: str,
    document_id: str,
    page_number: int,
    chunk_size: int | None = None,
    chunk_overlap: int | None = None,
) -> list[TextChunk]:
    """Split text into overlapping chunks using recursive boundary detection."""
    chunk_size = chunk_size or settings.chunk_size
    chunk_overlap = chunk_overlap or settings.chunk_overlap

    cleaned = clean_text(text)

    if len(cleaned) <= chunk_size:
        if not cleaned:
            return []
        return [
            TextChunk(
                text=cleaned,
                index=0,
                page_number=page_number,
                document_id=document_id,
                start_char=0,
                end_char=len(cleaned),
            )
        ]

    chunks: list[TextChunk] = []
    start = 0
    chunk_index = 0

    while start < len(cleaned):
        end = start + chunk_size

        if end < len(cleaned):
            # Try paragraph boundary
            para_break = cleaned.rfind("\n\n", start + int(chunk_size * 0.5), end)
            if para_break > start:
                end = para_break
            else:
                # Try sentence boundary
                sent_break = find_last_sentence_break(cleaned, start + int(chunk_size * 0.3), end)
                if sent_break > start:
                    end = sent_break
                else:
                    # Try word boundary
                    word_break = cleaned.rfind(" ", start, end)
                    if word_break > start:
                        end = word_break
        else:
            end = len(cleaned)

        chunk_text_str = cleaned[start:end].strip()
        if chunk_text_str:
            chunks.append(
                TextChunk(
                    text=chunk_text_str,
                    index=chunk_index,
                    page_number=page_number,
                    document_id=document_id,
                    start_char=start,
                    end_char=end,
                )
            )
            chunk_index += 1

        start = end - chunk_overlap
        if start >= len(cleaned):
            break

    return chunks


def chunk_document(
    pages: list[dict],
    document_id: str,
    chunk_size: int | None = None,
    chunk_overlap: int | None = None,
) -> list[TextChunk]:
    """Chunk a multi-page document."""
    all_chunks: list[TextChunk] = []
    global_index = 0

    for page in pages:
        page_chunks = chunk_text(
            page["text"],
            document_id,
            page["page_number"],
            chunk_size,
            chunk_overlap,
        )
        for chunk in page_chunks:
            chunk.index = global_index
            all_chunks.append(chunk)
            global_index += 1

    return all_chunks
